File: ts/project/.sdk/tm/csharp/test/NetsimTest.cs
// Network-behaviour simulation over the offline mock transport. The `test`
// feature accepts an optional `net` config so unit tests can exercise slow,
// failing and offline conditions without a live server. These checks drive
// the transport through Direct(), which needs no entity, so they run for
// every generated SDK regardless of its API shape.
// C# twin of tm/go/test/netsim_test.go.

using System.Diagnostics;

using Xunit;

using ProjectNameSdk;

namespace ProjectNameSdk.Test;

public class NetsimTest
{
    [Fact]
    public void OfflineSimulationFailsRequest()
    {
        var client = ProjectNameSDK.TestSDK(new Dictionary<string, object?>
        {
            ["net"] = new Dictionary<string, object?> { ["offline"] = true },
        }, null);
        var res = client.Direct(new Dictionary<string, object?> { ["path"] = "/ping" });
        Assert.True(Equals(res["ok"], false), $"offline network must fail the call");
    }

    [Fact]
    public void FailstatusSimulationSurfacesStatus()
    {
        var client = ProjectNameSDK.TestSDK(new Dictionary<string, object?>
        {
            ["net"] = new Dictionary<string, object?>
            {
                ["failTimes"] = 1,
                ["failStatus"] = 503,
            },
        }, null);
        var res = client.Direct(new Dictionary<string, object?> { ["path"] = "/ping" });
        Assert.True(Equals(res["ok"], false), "expected failed call");
        Assert.Equal(503, Helpers.ToInt(res["status"]));
    }

    [Fact]
    public void LatencySimulationDelaysRequest()
    {
        var delay = 60;
        var client = ProjectNameSDK.TestSDK(new Dictionary<string, object?>
        {
            ["net"] = new Dictionary<string, object?> { ["latency"] = delay },
        }, null);
        var sw = Stopwatch.StartNew();
        client.Direct(new Dictionary<string, object?> { ["path"] = "/ping" });
        sw.Stop();
        // Generous lower bound to stay robust on slow CI.
        Assert.True(sw.ElapsedMilliseconds >= delay - 25,
            $"expected >= {delay - 25}ms latency, got {sw.ElapsedMilliseconds}ms");
    }

    [Fact]
    public void PlainTestSdkWorksWithoutNet()
    {
        var client = ProjectNameSDK.TestSDK(null, null);
        Assert.NotNull(client);
    }
}

File: ts/project/.sdk/tm/go/test/netsim_test.go
package sdktest

// Network-behaviour simulation over the offline mock transport. The `test`
// feature accepts an optional `net` config so unit tests can exercise slow,
// failing and offline conditions without a live server. These checks drive
// the transport through Direct(), which needs no entity, so they run for
// every generated SDK regardless of its API shape.

import (
	"testing"
	"time"

	sdk "GOMODULE"
)

func TestNetsim(t *testing.T) {

	t.Run("offline-simulation-fails-request", func(t *testing.T) {
		client := sdk.TestSDK(map[string]any{
			"net": map[string]any{"offline": true},
		}, nil)
		res, err := client.Direct(map[string]any{"path": "/ping"})
		if err != nil {
			t.Fatalf("Direct itself should not error: %v", err)
		}
		if res["ok"] != false {
			t.Errorf("offline network must fail the call, got %v", res)
		}
	})

	t.Run("failstatus-simulation-surfaces-status", func(t *testing.T) {
		client := sdk.TestSDK(map[string]any{
			"net": map[string]any{"failTimes": 1, "failStatus": 503},
		}, nil)
		res, err := client.Direct(map[string]any{"path": "/ping"})
		if err != nil {
			t.Fatalf("Direct itself should not error: %v", err)
		}
		if res["ok"] != false {
			t.Errorf("expected failed call, got %v", res)
		}
		if res["status"] != 503 {
			t.Errorf("expected simulated 503, got %v", res["status"])
		}
	})

	t.Run("latency-simulation-delays-request", func(t *testing.T) {
		delay := 60
		client := sdk.TestSDK(map[string]any{
			"net": map[string]any{"latency": delay},
		}, nil)
		start := time.Now()
		_, err := client.Direct(map[string]any{"path": "/ping"})
		if err != nil {
			t.Fatalf("Direct itself should not error: %v", err)
		}
		elapsed := time.Since(start).Milliseconds()
		// Generous lower bound to stay robust on slow CI.
		if elapsed < int64(delay-25) {
			t.Errorf("expected >= %dms latency, got %dms", delay-25, elapsed)
		}
	})

	t.Run("plain-test-sdk-works-without-net", func(t *testing.T) {
		client := sdk.TestSDK(nil, nil)
		if client == nil {
			t.Fatal("expected a client")
		}
	})
}
